src/Interpreters/InterpreterInsertQuery.cpp

```diff
@@ -200,9 +200,47 @@ BlockIO InterpreterInsertQuery::execute()
         size_t out_streams_size = 1;
         if (query.select)
         {
-            /// Passing 1 as subquery_depth will disable limiting size of intermediate result.
-            InterpreterSelectWithUnionQuery interpreter_select{ query.select, context, SelectQueryOptions(QueryProcessingStage::Complete, 1)};
-            res = interpreter_select.execute();
+            auto is_trivial_select = [](const auto query_)
+            {
+                if (!query_.distinct && !query_.limit_with_ties && !query_.prewhere() && !query_.where() && !query_.groupBy()
```

Comment:
Also let's check that the table is not a subquery... or better - that if it's a subquery it's also trivial.